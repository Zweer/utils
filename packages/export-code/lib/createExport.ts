import { mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { dirname, extname } from 'node:path';

const defaultTemplate = `# EXPORT

## File structure

\`\`\`
{TREE}
\`\`\`

## File export

{FILES}
{FILENAME}

{CODE}
{NOTLAST}
---

{NOTLAST}
{FILES}`;

function getLanguageForExtension(extension: string): string {
  const langMap: Record<string, string> = {
    '.js': 'javascript',
    '.ts': 'typescript',
    '.jsx': 'jsx',
    '.tsx': 'tsx',
    '.json': 'json',
    '.md': 'markdown',
    '.html': 'html',
    '.css': 'css',
    '.scss': 'scss',
    '.yml': 'yaml',
    '.yaml': 'yaml',
    '.sh': 'shell',
    '.py': 'python',
  };
  return langMap[extension.toLowerCase()] || '';
}

export function generateExportFile(
  filenames: string[],
  tree: string,
  exportPath: string,
  template: string = defaultTemplate,
): void {
  const fileLoopRegex = /\{FILES\}\n(.*?)\n\{FILES\}/s;
  const loopMatch = template.match(fileLoopRegex);

  if (!loopMatch || !loopMatch[1]) {
    console.error('Template doesn\'t contain {FILES} section');
    throw new Error('InvalidTemplate');
  }
  const fileTemplate = loopMatch[1];

  const notLastRegex = /\{NOTLAST\}(.*?)\{NOTLAST\}/s;
  const notLastMatch = fileTemplate.match(notLastRegex);
  const separatorTemplate = notLastMatch ? notLastMatch[1] : '';

  const allFilesContent = filenames.map((filename, index) => {
    const isLastFile = index === filenames.length - 1;

    const fileContent = readFileSync(filename, 'utf8');
    const extension = extname(filename);
    const language = getLanguageForExtension(extension);
    const codeBlock = `\`\`\`\`${language}\n${fileContent}\n\`\`\`\``;
    const currentFileBlock = fileTemplate
      .replace('{FILENAME}', filename)
      .replace('{CODE}', codeBlock)
      .replace(notLastRegex, isLastFile ? '' : separatorTemplate);

    return currentFileBlock;
  });

  const finalOutput = template
    .replace('{TREE}', tree)
    .replace(fileLoopRegex, allFilesContent.join(''));

  const exportDirectory = dirname(exportPath);
  mkdirSync(exportDirectory, { recursive: true });

  writeFileSync(exportPath, finalOutput);
}
