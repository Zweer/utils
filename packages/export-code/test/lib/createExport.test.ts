import { vol } from 'memfs';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';

import { generateExportFile } from '../../lib/createExport.js';

vi.mock('node:fs', async () => {
  const memfs = await vi.importActual<typeof import('memfs')>('memfs');
  return memfs.fs;
});

describe('generateExportFile', () => {
  const MOCK_FILES_STRUCTURE = {
    '/project/src/index.ts': 'console.log("Hello, TypeScript!");',
    '/project/README.md': '# My Awesome Project',
    '/project/src/utils.js': 'const a = () => `hello ``` world`;',
  };
  const MOCK_FILENAMES = Object.keys(MOCK_FILES_STRUCTURE);
  const MOCK_TREE = `project
├── README.md
└── src
    ├── index.ts
    └── utils.js`;
  const EXPORT_PATH = '/project/export.md';

  beforeEach(() => {
    vol.fromJSON(MOCK_FILES_STRUCTURE, '/');
  });

  afterEach(() => {
    vol.reset();
  });

  it('should generate a file using the default template', () => {
    generateExportFile(MOCK_FILENAMES, MOCK_TREE, EXPORT_PATH);

    expect(vol.existsSync(EXPORT_PATH)).toBe(true);

    const output = vol.readFileSync(EXPORT_PATH, 'utf8') as string;

    expect(output).toMatchSnapshot();
  });

  it('should correctly apply a custom template', () => {
    const customTemplate = `MY CUSTOM EXPORT
Tree:
{TREE}

Files:
{FILES}
--- FILE: {FILENAME} ---
{CODE}
{NOTLAST}
*** SEPARATOR ***
{NOTLAST}
{FILES}
`;

    generateExportFile(MOCK_FILENAMES, MOCK_TREE, EXPORT_PATH, customTemplate);

    const output = vol.readFileSync(EXPORT_PATH, 'utf8') as string;

    expect(output).toContain('MY CUSTOM EXPORT');
    expect(output).toContain('*** SEPARATOR ***');
    expect(output).not.toContain('## File structure');

    expect(output).toMatchSnapshot();
  });

  it('should throw an error if the template is invalid (missing {FILES})', () => {
    const invalidTemplate = 'This template has no file section.';

    const action = () => generateExportFile([], MOCK_TREE, EXPORT_PATH, invalidTemplate);

    expect(action).toThrow('InvalidTemplate');
  });

  it('should not include the {NOTLAST} separator for the last file', () => {
    const template = `{FILES}{FILENAME}{NOTLAST}---{NOTLAST}{FILES}`;

    generateExportFile(MOCK_FILENAMES, MOCK_TREE, EXPORT_PATH, template);

    const output = vol.readFileSync(EXPORT_PATH, 'utf8') as string;

    const expectedOutput = MOCK_FILENAMES.join('---');

    expect(output).toBe(expectedOutput);
  });

  it('should handle an empty list of filenames gracefully', () => {
    generateExportFile([], MOCK_TREE, EXPORT_PATH);

    const output = vol.readFileSync(EXPORT_PATH, 'utf8') as string;

    expect(output).toContain(MOCK_TREE);
    expect(output).not.toContain('{FILENAME}');
    expect(output).not.toContain('{CODE}');

    expect(output).toMatchSnapshot();
  });

  it('should correctly handle a single file in the list (no separator)', () => {
    const singleFilename = ['/project/README.md'];
    generateExportFile(singleFilename, 'project\n└── README.md', EXPORT_PATH);

    const output = vol.readFileSync(EXPORT_PATH, 'utf8') as string;

    expect(output).not.toContain('\n---\n');
    expect(output).toContain('# My Awesome Project');
    expect(output).toMatchSnapshot();
  });

  it('should correctly create code blocks with language identifiers', () => {
    generateExportFile(MOCK_FILENAMES, MOCK_TREE, EXPORT_PATH);
    const output = vol.readFileSync(EXPORT_PATH, 'utf8') as string;

    expect(output).toContain('````typescript\nconsole.log("Hello, TypeScript!");');
    expect(output).toContain('````markdown\n# My Awesome Project');
    expect(output).toContain('````javascript\nconst a = () => `hello ``` world`;');
  });
});
