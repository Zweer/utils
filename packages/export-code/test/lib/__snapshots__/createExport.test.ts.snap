// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateExportFile > should correctly apply a custom template 1`] = `
"MY CUSTOM EXPORT
Tree:
project
├── README.md
└── src
    ├── index.ts
    └── utils.js

Files:

--- FILE: /project/src/index.ts ---
\`\`\`\`typescript
console.log("Hello, TypeScript!");
\`\`\`\`

*** SEPARATOR ***


--- FILE: /project/README.md ---
\`\`\`\`markdown
# My Awesome Project
\`\`\`\`

*** SEPARATOR ***


--- FILE: /project/src/utils.js ---
\`\`\`\`javascript
const a = () => \`hello \`\`\` world\`;
\`\`\`\`


"
`;

exports[`generateExportFile > should correctly handle a single file in the list (no separator) 1`] = `
"# EXPORT

## File structure

\`\`\`
project
└── README.md
\`\`\`

## File export


/project/README.md

\`\`\`\`markdown
# My Awesome Project
\`\`\`\`


"
`;

exports[`generateExportFile > should generate a file using the default template 1`] = `
"# EXPORT

## File structure

\`\`\`
project
├── README.md
└── src
    ├── index.ts
    └── utils.js
\`\`\`

## File export


/project/src/index.ts

\`\`\`\`typescript
console.log("Hello, TypeScript!");
\`\`\`\`



---


/project/README.md

\`\`\`\`markdown
# My Awesome Project
\`\`\`\`



---


/project/src/utils.js

\`\`\`\`javascript
const a = () => \`hello \`\`\` world\`;
\`\`\`\`


"
`;

exports[`generateExportFile > should handle an empty list of filenames gracefully 1`] = `
"# EXPORT

## File structure

\`\`\`
project
├── README.md
└── src
    ├── index.ts
    └── utils.js
\`\`\`

## File export

"
`;
